File: LeetCode/Python/Easy/234-Palindrome-Linked-List/solution.py
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head==None:
            return True

        if head.next==None:
            return True

        #Calculate the length of the list
        length: int=self.get_length(head)

        #Determine the mid point of the list
        middle: int=length // 2

        #Stack to hold the nodes in the first half of the list
        stack: List[ListNode]=list()

        current: ListNode=head

        #Iterate through the first half of the list
        for count in range(0, middle):
            
            #Add the node to the stack
            stack.append(current.val)

            #Move the pointer
            current=current.next

        #If the length of the list is odd, then after we iterate through the first half of the list, current will be pointing to the middle element.
        #So, we need to increment the pointer one position further in the list so it points to the first node in the second half of the list
        if length % 2 == 1:
            current=current.next

        #While current is not None (null) and there are still nodes in the stack
        while current!=None and len(stack) > 0:
            
            #Compare the node in the second half of the list with the current end node of the first half of the list
            #If the node values don't match, then the list is not a palindrome
            if current.val!=stack[-1]:
                return False
            
            #Move the pointer
            current=current.next

            #Remove the node from the stack
            stack.pop()

        return True

    def get_length(self, head: Optional[ListNode]) -> int:
        length: int=0

        current: ListNode=head

        while current!=None:
            length+=1

            current=current.next

        return length

File: LeetCode/Rust/Easy/234-Palindrome-Linked-List/solution.rs
impl Solution
{
    pub fn is_palindrome(head: Option<Box<ListNode>>) -> bool
    {
        let mut stack: Vec<i32>=Vec::new();

        let length: usize=Solution::get_length(head.as_ref());

        let middle: usize=length / 2;

        let mut current: Option<&Box<ListNode>>=head.as_ref();

        for counter in 0..middle
        {
            stack.push(current.unwrap().val);

            current=current.unwrap().next.as_ref();
        }

        if length % 2 == 1
        {
            current=current.unwrap().next.as_ref();
        }

        while let Some(node)=current
        {
            let top: i32=stack.pop().unwrap();

            if top!=node.val
            {
                return false;
            }

            current=node.next.as_ref();
        }

        true
    }

    pub fn get_length(head: Option<&Box<ListNode>>) -> usize
    {
        let mut length: usize=0;

        let mut current: Option<&Box<ListNode>>=head;

        while let Some(node)=current
        {
            length+=1;

            current=node.next.as_ref();
        }

        length
    }
}